go/libraries/doltcore/sqle/tables.go

```diff
@@ -400,22 +407,176 @@ func (t *DoltTable) DataLength(ctx *sql.Context) (uint64, error) {
 	return numBytesPerRow * numRows, nil
 }
 
-// TODO: Have actual implementations for AnalyzeTable and Statistics; this is just a quick fix to unblock others.
-
 // AnalyzeTable implements the sql.StatisticsTable interface.
 func (t *DoltTable) AnalyzeTable(ctx *sql.Context) error {
+	table, err := t.DoltTable(ctx)
+	if err != nil {
+		return err
+	}
+
+	m, err := table.GetRowData(ctx)
+	if err != nil {
+		return err
+	}
+
+	// initialize dolt stats
+	cols := t.sch.GetAllCols().GetColumns()
+	t.doltStats = &DoltTableStatistics{
+		rowCount:     m.Count(),
+		nullCount:    0,
+		createdAt:    time.Now(),
+		histogramMap: make(sql.HistogramMap),
+	}
+
+	// initialize histogram map
+	for _, col := range cols {
+		hist := new(sql.Histogram)
+		hist.Min = math.MaxFloat64
+		hist.Max = -math.MaxFloat64
+		t.doltStats.histogramMap[col.Name] = hist
+	}
+
+	// this can be adapted to a histogram with any number of buckets
+	freqMap := make(map[string]map[float64]uint64)
+	for _, col := range cols {
+		freqMap[col.Name] = make(map[float64]uint64)
+	}
+
+	//TODO: async?
+	partIter, err := t.Partitions(ctx)
+	if err != nil {
+		return err
+	}
+
+	for {
+		part, err := partIter.Next(ctx)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		iter, err := t.PartitionRows(ctx, part)
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return err
+		}
+
+		for {
+			row, err := iter.Next(ctx)
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return err
+			}
+
+			for i, col := range cols {
+				hist, ok := t.doltStats.histogramMap[col.Name]
+				if !ok {
+					panic("histogram was not initialized for this column; shouldn't be possible")
+				}
+
+				if row[i] == nil {
+					hist.NullCount++
+					continue
+				}
+
+				val, err := sql.Float64.Convert(row[i])
+				if err != nil {
+					continue // skip unsupported column types for now
+				}
+				v := val.(float64)
+
+				// place into frequency map
+				if bucket, ok := freqMap[col.Name][v]; ok {
+					bucket++
+				} else {
+					freqMap[col.Name][v] = 1
+					hist.DistinctCount++
+				}
+
+				hist.Mean += v
+				hist.Min = math.Min(hist.Min, v)
+				hist.Max = math.Max(hist.Max, v)
+				hist.Count++
+			}
+		}
+	}
+
+	// TODO: logic to determine ranges for buckets
```

Comment:
merged comments should either be: 1) consolidated into a docstring to describe what the function does, or 2) explain tricky edge cases that future readers would stumble on